provider/gateway/client.go

```diff
@@ -144,5 +158,79 @@ func makeURI(host string, path string) (string, error) {
 	if err != nil {
 		return "", err
 	}
+
 	return endpoint.String(), nil
 }
+
+func (c *client) ServiceLogs(ctx context.Context,
+	host string,
+	id mtypes.LeaseID,
+	service string,
+	follow bool,
+	tailLines int64) (*ServiceLogs, error) {
+
+	endpoint, err := url.Parse(host + "/" + serviceLogsPath(id, service))
+	if err != nil {
+		return nil, err
+	}
+
+	switch endpoint.Scheme {
+	case "ws", "http", "":
+		endpoint.Scheme = "ws"
+	case "wss", "https":
+		endpoint.Scheme = "wss"
+	default:
+		return nil, errors.Errorf("invalid uri scheme \"%s\"", endpoint.Scheme)
+	}
+
+	query := url.Values{}
+
+	query.Set("follow", strconv.FormatBool(follow))
+	query.Set("tail", strconv.FormatInt(tailLines, 10))
+
+	endpoint.RawQuery = query.Encode()
+
+	conn, _, err := websocket.DefaultDialer.DialContext(ctx, endpoint.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// todo check status?
+
+	streamch := make(chan ServiceLogMessage)
+	logs := &ServiceLogs{
+		Stream: streamch,
+	}
+
+	go func(conn *websocket.Conn) {
+		defer func() {
+			close(streamch)
+			_ = conn.Close()
+		}()
+
+		for {
+			_ = conn.SetReadDeadline(time.Now().Add(pingWait))
+			mType, msg, e := conn.ReadMessage()
+			if e != nil {
+				return
+			}
+
+			switch mType {
+			case websocket.PingMessage:
+				if e = conn.WriteMessage(websocket.PongMessage, []byte{}); e != nil {
+					return
+				}
+			case websocket.TextMessage:
+				var logLine ServiceLogMessage
+				if e = json.Unmarshal(msg, &logLine); e != nil {
+					return
+				}
+
+				streamch <- logLine
```

Comment:
:thinking: I wonder if we should handle the `websocket.CloseMessage`?  we can add it later if so.

Reply:
good point